File: lib/equation_parser.rb
require 'pry'
require 'contracts'

class EquationParser
  include Contracts::Core
  include Contracts::Builtin

  attr_accessor :left_hash

  Contract String => String
  def initialize(data)
    @left_hash = {}
    data.each_line do |line|
      split_equation = self.split_on_equal(line)
      @left_hash[split_equation[0]] = split_equation[1]
    end
  end

  Contract String => Array
  def split_on_equal(string)
    parsed_string = string.split('=')
    parsed_string.each do |item|
      item.strip!
    end
  end

  Contract String => Array
  def split_on_operator(string)
    parsed_string = string.split('+')
    parsed_string.each do |item|
      item.strip!
    end
  end

  Contract Or[String, Int] => Bool
  def is_integer(string)
    return true if string.is_a? Integer
    string =~ /^\d+$/ ? true : false
  end

  Contract String => Or[String, Int]
  def resolve_variable(string)
    @left_hash[string]
  end

  Contract Or[String, Int] => Or[String, Int]
  def compute(string)
    result = 0
    if is_integer(string)
      return string.to_i
    else 
      expanded_string = split_on_operator(string)
    end
    expanded_string.each do |item|
      if is_integer(item)
        result += item.to_i
      else 
        result += compute(resolve_variable(item))
      end
    end
    return result
  end

  Contract nil => String
  def evaluate
    return_value = []
    @left_hash.each do |key, value|
      result = 0
      operation_array = split_on_operator(value)
      operation_array.each do |operand|
        result += compute(operand)
      end
      @left_hash[key] = result
      return_value.push("#{key} = #{result}")
    end
    return_value.sort.join("\n")
  end

end

File: cpp/src/equationparser.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <fstream>
#include <iostream>
#include <vector>
#include <map>
#include <sstream>
#include <string>
#include <regex>
#include "../includes/equationparser.h"

using namespace std;

EquationParser::EquationParser(vector<string> lines) {
    for (auto line: lines) {
        vector<string> split_equation = splitOnChar(line, '=');
        left_hash[split_equation[0]] = split_equation[1];
    }
}

void EquationParser::updateMap(string key, string value) {
    left_hash[key] = value;
}

void EquationParser::dumpMap() {
    cout << "dumping map..." << endl;
    for (auto item : left_hash) {
        cout << item.first << " => " << item.second << endl;
    }
}

vector<string> EquationParser::splitOnChar(string line, char character) {
    stringstream ss_line(line);
    string segment;
    vector<string> seglist;

    while(getline(ss_line, segment, character)) {
        seglist.push_back(trim(segment));
    }
    stringstream output;
    output << "[";
    for (auto item: seglist) {
        output << item << ",";
    }
    return seglist;
}

string EquationParser::trim(string rawString) {
    if(rawString.empty()) {
        return rawString;
    }

    string left_trim = regex_replace(rawString, regex("^\\s+\\b"), "");
    string right_trim = regex_replace(left_trim, regex("\\b\\s+$"), "");

    return right_trim;
}

bool EquationParser::isInteger(string input) {
    regex re ("^\\d+$");
    if (regex_match(input, re)) return true;
    else return false;
}

bool EquationParser::containsOperator(string input) {
    regex re ("^.*\\++.*$");
    if (regex_match(input, re)) return true;
    else return false;
}

string EquationParser::resolveVariable(string query) {
    return left_hash[query];
}

string EquationParser::compute(string input) {
    int result = 0;
    vector<string> expanded_string;

    if (isInteger(input)) {
        return input;
    } else if (containsOperator(input)) {
        expanded_string = splitOnChar(input, '+');
        for (auto item: expanded_string) {
            if (isInteger(item)) {
                result += stoi(item);
            } else {
                result += stoi(compute(resolveVariable(item)));
            }

        }
    } else {
        result += stoi(compute(resolveVariable(input)));
    }
    return to_string(result);
}

string EquationParser::evaluate() {
    vector<string> operation_array;
    stringstream formatted_return;

    for (auto item: left_hash) {
        int result = 0;
        stringstream return_line;
        operation_array = splitOnChar(item.second, '+');
        for(auto operand: operation_array) {
            result += stoi(compute(operand));
        }
        updateMap(item.first, to_string(result));
        return_line << item.first << " = " << result << endl;
        formatted_return << return_line.str();
    }
    return formatted_return.str();
}
